### ISSUE TITLE
SagaAudit Duplication Key Fatal Exception, but with no information on what key

### ISSUE THREAD

**Issue by danielHalan** [Particular/ServiceControl#977]
Hi, 

When trying to start Service Control a Fatal Exception is thrown causing Service Control to Stop, the error message informs that there is a duplicate Saga key, but not which ID. (Using Log Level = Trace)

Also how would one solve this issue? That is in which Raven collection or ASB queue would one find this duplicate record, that needs to be removed. 

``
2017-06-21 09:25:21.9356|50|Info|NServiceBus.Unicast.Transport.TransportReceiver|Failed to process message
System.ArgumentException: An item with the same key has already been added.
   at System.ThrowHelper.ThrowArgumentException(ExceptionResource resource)
   at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
   at ServiceControl.SagaAudit.SagaAuditing.SagaRelationshipsEnricher.Enrich(IReadOnlyDictionary`2 headers, IDictionary`2 metadata) in C:\BuildAgent\work\7189a56f9f44affc\src\ServiceControl\SagaAudit\SagaAuditing.cs:line 36
   at ServiceControl.Operations.AuditQueueImport.ConvertToSaveMessage(TransportMessage message) in C:\BuildAgent\work\7189a56f9f44affc\src\ServiceControl\Operations\AuditQueueImport.cs:line 118
   at ServiceControl.Operations.AuditQueueImport.InnerHandle(TransportMessage message) in C:\BuildAgent\work\7189a56f9f44affc\src\ServiceControl\Operations\AuditQueueImport.cs:line 93
   at ServiceControl.Operations.AuditQueueImport.Handle(TransportMessage message) in C:\BuildAgent\work\7189a56f9f44affc\src\ServiceControl\Operations\AuditQueueImport.cs:line 52
   at NServiceBus.Satellites.SatelliteLauncher.HandleMessageReceived(Object sender, TransportMessageReceivedEventArgs e, ISatellite satellite) in C:\BuildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Satellites\SatelliteLauncher.cs:line 89
   at NServiceBus.Unicast.Transport.TransportReceiver.OnTransportMessageReceived(TransportMessage msg) in C:\BuildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Unicast\Transport\TransportReceiver.cs:line 413
   at NServiceBus.Unicast.Transport.TransportReceiver.ProcessMessage(TransportMessage message) in C:\BuildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Unicast\Transport\TransportReceiver.cs:line 344
   at NServiceBus.Unicast.Transport.TransportReceiver.TryProcess(TransportMessage message) in C:\BuildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Unicast\Transport\TransportReceiver.cs:line 230
   at NServiceBus.Azure.Transports.WindowsAzureServiceBus.AzureServiceBusDequeueStrategy.TryProcessMessage(Object obj) in C:\BuildAgent\work\2f57832e2eee436e\src\Transport\Receiving\AzureServiceBusDequeueStrategy.cs:line 137
``

**Comment by danielHalan** [comment]
Also that error is finally followed by mutliple Fatal error, 

``
2017-06-21 09:26:05.2240|46|Fatal|NServiceBus|Failed to import too many times
System.ArgumentException: An item with the same key has already been added.
   at System.ThrowHelper.ThrowArgumentException(ExceptionResource resource)
   at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
   at ServiceControl.SagaAudit.SagaAuditing.SagaRelationshipsEnricher.Enrich(IReadOnlyDictionary`2 headers, IDictionary`2 metadata) in C:\BuildAgent\work\7189a56f9f44affc\src\ServiceControl\SagaAudit\SagaAuditing.cs:line 36
   at ServiceControl.Operations.AuditQueueImport.ConvertToSaveMessage(TransportMessage message) in C:\BuildAgent\work\7189a56f9f44affc\src\ServiceControl\Operations\AuditQueueImport.cs:line 118
   at ServiceControl.Operations.AuditQueueImport.InnerHandle(TransportMessage message) in C:\BuildAgent\work\7189a56f9f44affc\src\ServiceControl\Operations\AuditQueueImport.cs:line 93
   at ServiceControl.Operations.AuditQueueImport.Handle(TransportMessage message) in C:\BuildAgent\work\7189a56f9f44affc\src\ServiceControl\Operations\AuditQueueImport.cs:line 52
   at NServiceBus.Satellites.SatelliteLauncher.HandleMessageReceived(Object sender, TransportMessageReceivedEventArgs e, ISatellite satellite) in C:\BuildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Satellites\SatelliteLauncher.cs:line 89
   at NServiceBus.Unicast.Transport.TransportReceiver.OnTransportMessageReceived(TransportMessage msg) in C:\BuildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Unicast\Transport\TransportReceiver.cs:line 413
   at NServiceBus.Unicast.Transport.TransportReceiver.ProcessMessage(TransportMessage message) in C:\BuildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Unicast\Transport\TransportReceiver.cs:line 344
   at NServiceBus.Unicast.Transport.TransportReceiver.TryProcess(TransportMessage message) in C:\BuildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Unicast\Transport\TransportReceiver.cs:line 230
   at NServiceBus.Azure.Transports.WindowsAzureServiceBus.AzureServiceBusDequeueStrategy.TryProcessMessage(Object obj) in C:\BuildAgent\work\2f57832e2eee436e\src\Transport\Receiving\AzureServiceBusDequeueStrategy.cs:line 137
``

**Comment by SzymonPobiega** [comment]
@danielHalan Which version(s) of NServiceBus do you use? It seems like the problem is caused by a message which invoked the same saga instance more the once. I'll check in what cases this is possible. 

The messages that failed should be logged in `FailedImports\Audit` folder and also stored in `FailedAuditImport` collection in RavenDB. When SC restarts, it will attempt to re-import messages from that collection.

Can you send the headers of one of the failed messages to our support e-mail address or open a case and attach them? This way we can try to figure out which saga is causing the problem and why.

**Comment by SzymonPobiega** [comment]
Indeed I managed to confirm that if a saga is defined like this:
```
class MySaga : Saga<MySagaData>, IAmStartedByMessages<MyMessage>, IHandleMessages<IMessage>
```

when the saga persistence is non-transactional it generates the audit message containing a header that causes SC to blow up:

```
Key: ServiceControl.SagaStateChange
Value: 51b5ad68-8ac4-46ee-a39c-a79900ca4ea8:New;51b5ad68-8ac4-46ee-a39c-a79900ca4ea8:Updated
```

**Comment by SzymonPobiega** [comment]
@danielHalan can you confirm you have a similar saga? It is enough to have handlers for two message types that are in the same inheritence hierarchy (in my case `MyMessage` inherits from `IMessage`).

**Comment by SzymonPobiega** [comment]
@danielHalan before we publish a fix, could you disable the saga audit plugin in the endpoint that is hosting the problematic saga?

**Comment by SzymonPobiega** [comment]
There is a related bug in the Saga Audit plugin itself (endpoint side) which generates malformed InvokedSagas headers: https://github.com/Particular/ServiceControl.Plugin.Nsb6.SagaAudit/issues/24

**Comment by SzymonPobiega** [comment]
A PR fixing this: https://github.com/Particular/ServiceControl/pull/978

**Comment by danielHalan** [comment]
Actually all sagas should have unique Messages, i.e. not IHandle same message twice or a subclass/interface, but they all have same base class.

Is there any way to figure out which is the problematic Saga?

Some background regarding the issue, there has not been any new Sagas created or changes, since this stopped working, so my first assumption was that it might have been some data corruption. 
Regarding FailedAuditImports collection, it has messages from 1+yr so the amount is 100k+ 

Using NSB 5.2.21, with NSB.Azure 6.2.3, ServiceControl.Plugin.Nsb5.SagaAudit 1.0.2

**Comment by SzymonPobiega** [comment]
@danielHalan the failed messages contain "NServiceBus.InvokedSagas" header which contains ID and types of the invoked sagas. We need to look at them.

I think you didn't notice the problem before because the message arrival pattern ensured that there are some non-saga (correct) audit messages between the saga messages. The circuit breaker was reset every now and then. What has happened now is you got a streak of saga messages with invalid headers and that caused SC to fall over.

**Comment by danielHalan** [comment]
@SzymonPobiega Ok, can see the two sagas that it reefers to, and they use same type for IHandleTimeouts<>, based on earlier suggestion, one should use diff Message Type's for each saga to resolve issue?

Also to get Service Control to start again, should one empty/rename the FailedAuditImport-collection in RavenDB?

**Comment by SzymonPobiega** [comment]
@danielHalan we're working on a fix ( #978 ) that will allow you to upgrade SC and run just fine. We should be able to release it tomorrow.

Sharing the same message (including timeouts) types between sagas is supported. But the thing I wonder is why is a given instance of a timeout routed to two saga instances. The timeout should only be routed to the instance that requested it.

**Comment by SzymonPobiega** [comment]
@danielHalan could you send us the code for the sagas via email (to the support address)? It seems weird that a timeout message is routed to two instances. The fix we're release for SC will solve your message processing problem but there might be other issue related to the timeout handling.

**Comment by danielHalan** [comment]
@SzymonPobiega Great, will verify as soon its released.

Regarding the Timeout message, The concept was that there is one base-class for two sagas that implement the shared Timeout(msg) method, and that base class had the IHandle<> assignments, possible this that caused the issue.  Have refactored so each saga has own message type and just passes a shared interface to base class, lets see if it works after SC is back up.

**Comment by SzymonPobiega** [comment]
@danielHalan was this timeout mapped as starting a saga?

**Comment by danielHalan** [comment]
@SzymonPobiega no just regular IHandleTimeouts<> mapping

**Comment by SzymonPobiega** [comment]
@danielHalan I've tried something like this:

```
class MySaga1 : MyBaseSaga {}
class MySaga2 : MyBaseSaga {}

abstract class MyBaseSaga : Saga<MyBaseSaga.MySagaData>,
    IAmStartedByMessages<MyMessage>,
    IHandleTimeouts<MyTimeout>
{

    public Task Handle(MyMessage message, IMessageHandlerContext context)
    {
        Console.WriteLine("MyMessage handler");
        return RequestTimeout(context, DateTime.Now.AddSeconds(5), new MyTimeout());
    }
        
    protected override void ConfigureHowToFindSaga(SagaPropertyMapper<MyBaseSaga.MySagaData> mapper)
    {
        mapper.ConfigureMapping<MyMessage>(m => m.Unique).ToSaga(s => s.Unique);
    }
        
    public class MySagaData : ContainSagaData
    {
        public Guid Unique { get; set; }
    }

    public Task Timeout(MyTimeout state, IMessageHandlerContext context)
    {
        Console.WriteLine("Timeout");
        return Task.FromResult(0);
    }
}
```

But it worked correctly (each timeout was routed only to its origin saga instance).

**Comment by danielHalan** [comment]
@SzymonPobiega the only diff. I can see is that the interface IHandleTimeouts was set on the top classes, MySaga1 and MySaga2.

**Comment by SzymonPobiega** [comment]
@danielHalan I think I found the issue. It is the fact that the saga data type is shared cause the problem. Because the lookups in `ISagaPersister` implementations are based on the saga data type, not the saga class type, it cannot differentiate between `MySaga1` and `MySaga2`.

In my repro the result is that there is a single saga data object created and the audited `MyMessage` contains both sagas in the `InvokedSagas`:

```
ConsoleApp1.MySaga1:3092188a-e64d-48ae-9ce5-a79b00a6af50ConsoleApp1.MySaga1:3092188a-e64d-48ae-9ce5-a79b00a6af50;ConsoleApp1.MySaga2:3092188a-e64d-48ae-9ce5-a79b00a6af50
```
Notice both have the same ID.

Long story short, you need to split the sagas so that each has its own saga data type. I'll make sure as part of fixing the bug we will update the documentation and validation to prevent such designs.

**Comment by SzymonPobiega** [comment]
@danielHalan I think it would be a good idea to go though the saga data in the store and verify it is not corrupted. In my repro both saga instances shared the same data object. The messages (timeouts in this case) send downstream contained the same saga ID. So from the perspective of other components, there are two sagas but they both refer to same data.

If you want, we can schedule a call next week to review it together.

**Comment by danielHalan** [comment]
@SzymonPobiega Ok, all sagas related to this issue are 'completed' and could be discarded in Audit purpose. The Sagas has been corrected so both Sagas has own Timeout & Data-classes. What would be the next step to get SC up n running?

We could schedule a review of the Failed Audit messages, in case there are some other issues (as there was a few of them), you can drop me a line at my Github-handle (at) gmail.

**Comment by danielHalan** [comment]
SC v1.41 Service now starts properly. :+1:

After the latest version has finished processing, went back to see whats left in FailedAuditImports, and there are quite few messages there, not related to the discussed issue, and some are not related to Saga's. First question, what is the consequence of messages ending up in this collection? and second, where can one find more error information, as I can't see any at first glance in the documents themselves.

**Comment by SzymonPobiega** [comment]
@danielHalan good to hear it starts 🎉 

Regarding `FailedAuditImports` there is no consequence other than the fact that the messages that ended up there are not accessible via ServiceInsight. 

In the normal SC log path there should be a subfolder called `FailedImports\Audit`. In that folder there should be a separate `txt` file for each messaged that failed. The file contains the actual exception information.

**Comment by mikeminutillo** [comment]
@danielHalan @SzymonPobiega I removed this issue from the 1.41.1 milestone so that I can release without closing this issue. The fix has still been merged and will be released as a part of 1.41.1.

**Comment by SzymonPobiega** [comment]
@danielHalan did you manage to find the payloads of messages that failed? Is it all good now?

**Comment by danielHalan** [comment]
@SzymonPobiega it all good +1
